### PR TITLE
Add NucleiScorer

### PR DESCRIPTION
**Pull request by alessiamarcolini**
**Example**
```python
from PIL import Image

from histolab.scorer import NucleiScorer
from histolab.tile import Tile

im = Image.open("image.png")
tile = Tile(im, None, 0)
ns = NucleiScorer()
score = ns(tile)

```

### CODE REVIEW

File: src/histolab/filters/compositions.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        filters = imf.Compose(
+            [
+                imf.RgbToGrayscale(),
+                imf.OtsuThreshold(),
+                mof.BinaryDilation(),
+                mof.RemoveSmallHoles(),
+                mof.RemoveSmallObjects(),
+            ]
+        )
+        return filters
```
Comment (ernestoarbitrio) [inline src/histolab/filters/compositions.py:71]: not mandatory but we can directly return the whole statement avoiding a pointless assignment
```
return imf.Compose(
            [
                imf.RgbToGrayscale(),
                imf.OtsuThreshold(),
                mof.BinaryDilation(),
                mof.RemoveSmallHoles(),
                mof.RemoveSmallObjects(),
            ]
        )
```

File: src/histolab/filters/compositions.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        filters = imf.Compose(
+            [
+                imf.RgbToGrayscale(),
+                imf.OtsuThreshold(),
+                mof.BinaryDilation(),
+                mof.BinaryFillHoles(structure=np.ones((5, 5))),
+            ]
+        )
+        return filters
```
Comment (ernestoarbitrio) [inline src/histolab/filters/compositions.py:92]: same herfe

File: src/histolab/tile.py [hunk 3]
```diff
@@ -170,7 +189,7 @@ def _has_tissue_more_than_percent(self, tissue_percent: float = 80.0) -> bool:
             True if tissue represent more than ``tissue_percent`` % of the image, False
             otherwise.
         """
-        filters = self._enough_tissue_mask_filters
+        filters = FiltersComposition(Tile).tissue_mask_filters
```
Comment (ernestoarbitrio) [inline src/histolab/tile.py:192]: thif funcion `_has_tissue_more_than_percent` should be sorted in alphabetical order

File: tests/integration/test_scorer.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            (TILES.NO_TISSUE, 0.015577231690474633),
+        ),
+    )
+    def it_knows_nuclei_score(self, tile_img, expected_score):
+        tile = Tile(tile_img, None)
+        nuclei_scorer = scorer.NucleiScorer()
+
+        score = nuclei_scorer(tile, 0.6)
+
+        assert score == expected_score
```
Comment (ernestoarbitrio) [inline tests/integration/test_scorer.py:28]: IMHO this integration test should have more edge cases .... more levels and other weird slide example we can encounter in the `real life` 😄

File: tests/unit/filters/test_util.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+def test_mask_difference():
+    expected_mask_difference = np.array(
+        [[False, True, False, False], [False, False, False, False]]
+    )
+
+    mask_difference_ = mask_difference(BASE_MASK, BASE_MASK2)
+
+    assert type(mask_difference_) == np.ndarray
+    assert mask_difference_.dtype == "bool"
+    np.testing.assert_equal(mask_difference_, expected_mask_difference)
```
Comment (ernestoarbitrio) [inline tests/unit/filters/test_util.py:17]: I would use also fixture with dim > 1

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
Some minor comments + in general I would include more integration tests for the nuclei scorer. I'd like to cover much corner cases as possible.

File: tests/unit/filters/test_util.py (hunk 5 above)
Comment (alessiamarcolini) [reply tests/unit/filters/test_util.py:17]: it's actually 2d :)

File: src/histolab/tile.py [hunk 6]
```diff
@@ -29,6 +29,30 @@ def __init__(self, image: PIL.Image.Image, coords: CoordinatePair, level: int =
...
+            Filter or composition of filters to be applied
+
+        Returns
+        -------
+        Tile
+            Tile with the filters applied
+        """
+        filtered_image = filters(self.image)
+        if isinstance(filtered_image, np.ndarray):
+            filtered_image = PIL.Image.fromarray(filtered_image)
```
Comment (ernestoarbitrio) [inline src/histolab/tile.py:53]: in which condition this could be a `ndarray`?
I mean, if self.image is a known property and we apply a filter composition to it, the result type should be of 1 type

me == confused

File: src/histolab/scorer.py [hunk 7]
```diff
@@ -18,5 +22,61 @@ def __call__(self, tile: Tile) -> float:
...
+        ----------
+        tile : Tile
+            The tile to calculate the score from.
+
+        Returns
+        -------
+        float
+            Nuclei score
+        """
+        tissue_ratio = tile.tissue_ratio
```
Comment (ernestoarbitrio) [inline src/histolab/scorer.py:67]: I'd avoid this statement

just return: `nuclei_ratio * np.tanh(tile.tissue_ratio)`

File: src/histolab/filters/image_filters.py [hunk 8]
```diff
@@ -937,3 +938,33 @@ def __call__(self, img: PIL.Image.Image) -> PIL.Image.Image:
 
     def __repr__(self) -> str:
         return self.__class__.__name__ + "()"
+
+
+class YenThreshold(object):
+    """Mask image based on pixel above Yen threshold.
+
+    Compute Yen threshold on image and return boolean mask based on pixels above this
```
Comment (nicolebussola) [inline src/histolab/filters/image_filters.py:946]: below

File: src/histolab/filters/image_filters.py [hunk 9]
```diff
@@ -937,3 +938,33 @@ def __call__(self, img: PIL.Image.Image) -> PIL.Image.Image:
...
+    img : PIL.Image.Image
+        Input image.
+    relate : operator, optional
+        Operator to be used to compute the mask from the threshold. Default is
+        operator.lt
+
+    Returns
+    -------
+    np.ndarray
+        Boolean NumPy array where True represents a pixel above Yen's threshold.
```
Comment (nicolebussola) [inline src/histolab/filters/image_filters.py:960]: below

File: src/histolab/tiler.py [hunk 10]
```diff
@@ -578,28 +582,39 @@ def _highest_score_tiles(self, slide: Slide) -> List[Tuple[float, CoordinatePair
         List[Tuple[float, CoordinatePair]]
             List of tuples containing the score and the extraction coordinates for the
             tiles with the highest score. Each tuple represents a tile.
+        List[Tuple[float, CoordinatePair]]
+            List of tuples containing the scaled score between 0 and 1 and the
+            extraction coordinates for the tiles with the highest score. Each tuple
+            represents a tile.
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:588]: ??? 2 lists?

File: src/histolab/filters/image_filters_functional.py [hunk 11]
```diff
@@ -550,10 +550,10 @@ def otsu_threshold(
 ) -> np.ndarray:
     """Mask image based on pixel above Otsu threshold.
 
-    Compute Otsu threshold on image as a NumPy array and return boolean mask
-    based on pixels above this threshold.
+    Compute Otsu threshold on image and return boolean mask based on pixels above this
```
Comment (nicolebussola) [inline src/histolab/filters/image_filters_functional.py:553]: below

Comment (nicolebussola) [reply src/histolab/filters/image_filters_functional.py:553]: returns: ....below

File: src/histolab/filters/image_filters_functional.py [hunk 12]
```diff
@@ -870,3 +870,29 @@ def pen_marks(img: PIL.Image.Image) -> np.ndarray:
...
+    img : PIL.Image.Image
+        Input image.
+    relate : operator, optional
+        Operator to be used to compute the mask from the threshold. Default is
+        operator.lt
+
+    Returns
+    -------
+    np.ndarray
+        Boolean NumPy array where True represents a pixel above Yen's threshold.
```
Comment (nicolebussola) [inline src/histolab/filters/image_filters_functional.py:894]: below

File: src/histolab/filters/morphological_filters.py [hunk 13]
```diff
@@ -253,3 +253,27 @@ def __call__(self, np_mask) -> np.ndarray:
...
     def __repr__(self):
         return self.__class__.__name__ + "()"
+
+
+class WhiteTopHat(object):
+    """Return white top hat of an image.
+
+    The white top hat of an image is defined as the image minus its morphological
+    opening. This operation returns the bright spots of the image that are smaller than
+    the structuring element.
```
Comment (nicolebussola) [inline src/histolab/filters/morphological_filters.py:263]: The white top hat of an image is defined as the image minus its morphological
    opening with respect to a structuring element. This operation returns the bright spots of the image that are smaller than
    the structuring element.

File: src/histolab/tiler.py [hunk 14]
```diff
@@ -578,28 +582,39 @@ def _highest_score_tiles(self, slide: Slide) -> List[Tuple[float, CoordinatePair
...
+            scaled_scores, key=lambda x: x[0], reverse=True
+        )
         if self.n_tiles > 0:
             highest_score_tiles = sorted_tiles_by_score[: self.n_tiles]
+            highest_scaled_score_tiles = sorted_tiles_by_scaled_score[: self.n_tiles]
         elif self.n_tiles == 0:
             highest_score_tiles = sorted_tiles_by_score
+            highest_scaled_score_tiles = sorted_tiles_by_scaled_score
         else:
             raise ValueError(f"'n_tiles' cannot be negative ({self.n_tiles})")
```
Comment (ernestoarbitrio) [inline src/histolab/tiler.py:610]: I would change the order of this business logic:
```
if self.n_tiles < 0:
    raise ValueError(f"'n_tiles' cannot be negative ({self.n_tiles})")

if self._ntiles > 0:
    highest_score_tiles = sorted_tiles_by_score[: self.n_tiles]
    highest_scaled_score_tiles = sorted_tiles_by_scaled_score[: self.n_tiles]
else:
    highest_score_tiles = sorted_tiles_by_score
    highest_scaled_score_tiles = sorted_tiles_by_scaled_score
```

File: src/histolab/scorer.py [hunk 15]
```diff
@@ -18,5 +22,61 @@ def __call__(self, tile: Tile) -> float:
...
+        float
+            Random score ranging between 0 and 1.
+        """
         return np.random.random()
+
+
+class NucleiScorer(Scorer):
+    r"""Implement a Scorer that estimates the presence of nuclei in an H&E-stained tile.
+
+    An higher presence of nuclei is associated with an higher scorer, following this
```
Comment (nicolebussola) [inline src/histolab/scorer.py:46]: with a higher score.

**Review by nicolebussola — CHANGES_REQUESTED** [review]
minor changes wrt docstrings

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
left some comments :D ...

File: src/histolab/tile.py (hunk 6 above)
Comment (alessiamarcolini) [reply src/histolab/tile.py:53]: It depends on the transformations: if at the end we have an OtsuThreshold then the result will be a numpy array

File: src/histolab/tiler.py (hunk 10 above)
Comment (alessiamarcolini) [reply src/histolab/tiler.py:588]: yes because now we have to consider also the scaled version of the scores (from 0 to 1)

**Review by nicolebussola — APPROVED** [review]
LGTM

File: src/histolab/tile.py (hunk 6 above)
Comment (ernestoarbitrio) [reply src/histolab/tile.py:53]: got it

**Review by ernestoarbitrio — APPROVED** [review]
